File: fotoapparat-adapters/rxjava/src/main/java/io/fotoapparat/result/adapter/rxjava/ObservableAdapter.java
package io.fotoapparat.result.adapter.rxjava;

import java.util.concurrent.Future;

import io.fotoapparat.result.adapter.Adapter;
import rx.Observable;

/**
 * Adapter for {@link Observable}.
 */
public class ObservableAdapter<T> implements Adapter<T, Observable<T>> {

    private ObservableAdapter() {
    }

    /**
     * @return {@link Adapter} which adapts result to {@link Observable}.
     */
    public static <R> ObservableAdapter<R> toObservable() {
        return new ObservableAdapter<>();
    }

    @Override
    public Observable<T> adapt(Future<T> future) {
        return Observable.from(future);
    }
}

File: fotoapparat-adapters/rxjava/src/main/java/io/fotoapparat/result/adapter/rxjava/ObservableAdapter.kt
package io.fotoapparat.result.adapter.rxjava

import io.fotoapparat.result.PendingResult
import rx.Observable
import java.util.concurrent.Future

/**
 * Adapter for [Observable].
 */
object ObservableAdapter {

    /**
     * @return Adapter which adapts result to [Observable].
     */
    @JvmStatic
    fun <T> toObservable(): Function1<Future<T>, Observable<T>> {
        return { future -> Observable.from(future) }
    }

}

/**
 * @return A [Observable] from the given [PendingResult].
 */
fun <T> PendingResult<T>.toObservable(): Observable<T> {
    return adapt { future -> Observable.from(future) }
}
